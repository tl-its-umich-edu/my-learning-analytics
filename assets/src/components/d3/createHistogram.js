import * as d3 from 'd3'
import { adjustViewport } from '../../util/chart'
import { roundToXDecimals } from '../../util/math'
import { siteTheme } from '../../globals'

function createHistogram ({
  data, width, height, domElement, xAxisLabel, yAxisLabel, myGrade, maxGrade = 100,
  showNumberOnBars = false, showDashedLine = true
}) {
  const margin = { top: 20, right: 20, bottom: 50, left: 40 }
  const barColor = siteTheme.palette.secondary.main
  const [aWidth, aHeight] = adjustViewport(width, height, margin)

  // data usually will be [50.6, 50.6, 50.6, 50.6, 50.6, 74.28, 74.52, 75.89, 76.69,,.,.,.,.] lowest grades binned
  //  to hide low performers.

  // the set operation removes duplicates
  const tempUniqData = [...new Set(data)]
  const firstGradeAfterBinnedGrade = tempUniqData[1]

  /* only showing the tick values above the binned grades eg, with above data the tick will start from 75%
  * if course has 0% or > 95% grades then we just show the whole distribution
  * */
  const minTickGrade = showDashedLine ? Math.round(firstGradeAfterBinnedGrade / 5) * 5 : 0

  const x = d3.scaleLinear()
    .domain([0, maxGrade]).nice()
    .range([margin.left, aWidth - margin.right])

  const bins = d3.histogram()
    .domain(x.domain())
    .thresholds(x.ticks(40))(data)

  // getting the first bin that has some grades in them, accessing the x1(higher bin) value
  const dashLine = () => {
    for (const bin in bins) {
      if (bins[bin].length > 0) {
        return bins[bin].x1
      }
    }
  }

  const y = d3.scaleLinear()
    .domain([0, d3.max(bins, d => d.length)]).nice()
    .range([aHeight - margin.bottom, margin.top])

  const svg = d3.select(domElement).append('svg')
    .attr('width', aWidth)
    .attr('height', aHeight)

  const bar = svg.selectAll('rect')
    .data(bins).enter()
    .append('g')

  bar.append('rect')
    .attr('x', d => x(d.x0) + 1)
    .attr('width', d => Math.max(0, x(d.x1) - x(d.x0) - 1))
    .attr('y', d => y(d.length))
    .attr('height', d => y(0) - y(d.length))
    .attr('fill', barColor)

  // Default is not show count on bar text, in case of AB testing a feature we want to leave it a configurable option
  if (showNumberOnBars) {
    bar.append('text')
      .attr('x', d => x((d.x1 + d.x0) / 2))
      .attr('y', d => y(d.length) + margin.top - 5)
      .attr('text-anchor', 'middle')
      .attr('fill', 'white')
      .text(d => d.length === 0 ? '' : d.length)
  }

  const xAxis = g => g
    .attr('transform', `translate(0, ${aHeight - margin.bottom})`)
    .call(d3
      .axisBottom(x)
      .tickSizeOuter(0)
      .ticks(width > 750 ? 20 : 10)
      .tickFormat(d => `${d}%`)
    )
    .call(g => g.append('text')
      .attr('x', aWidth / 2)
      .attr('y', 40)
      .attr('fill', 'rgba(0, 0, 0, 0.87)')
      .attr('font-size', '0.875rem')
      .attr('font-weight', '400')
      .attr('line-height', '1.46429em')
      .text(xAxisLabel).attr('dy', -4)
    )
    .call(g => g.selectAll('.tick').filter(d => d < minTickGrade).remove())

  const yAxis = g => g
    .attr('transform', `translate(${margin.left},0)`)
    .attr('class', 'axis')
    .call(d3.axisLeft(y).tickSizeInner(-aWidth).ticks(5))
    .call(g => g.select('.domain').remove())
    .call(g => g.select('.tick:last-of-type text').clone()
      .attr('x', 4)
      .attr('fill', 'rgba(0, 0, 0, 0.87)')
      .attr('text-anchor', 'end')
      .attr('font-size', '0.875rem')
      .attr('font-weight', '400')
      .attr('line-height', '1.46429em')
      .attr('opacity', 1)
      .attr('text-anchor', 'start')
      .text(yAxisLabel).attr('dy', -4))

  svg.append('g')
    .call(xAxis)

  svg.append('g')
    .call(yAxis)

  if (myGrade) {
    svg.append('line')
      .attr('transform', `translate(0, ${aHeight - margin.bottom})`)
      .attr('x1', x(myGrade))
      .attr('y1', -aHeight + 75)
      .attr('x2', x(myGrade))
      .attr('y2', 0)
      .attr('stroke', 'darkorange')
      .attr('stroke-width', '2')
    svg.append('text')
      .attr('x', myGrade < 5 ? x(myGrade) + 3 : x(myGrade) - 120)
      .attr('y', margin.top + 15)
      .text(`My Grade: ${roundToXDecimals(myGrade, 1)}%`)
      .attr('font-size', '0.875rem')
      .attr('font-weight', 'bold')
      .attr('line-height', '1.46429em')
      .attr('text-anchor', 'start')
  }
  // Dashed line to show difference b/w binned and normal distribution
  if (showDashedLine) {
    svg.append('line')
      .attr('transform', `translate(0, ${aHeight - margin.bottom})`)
      .attr('x1', x(dashLine()))
      .attr('y1', -aHeight)
      .attr('x2', x(dashLine()))
      .attr('y2', 0)
      .attr('stroke-dasharray', ('3, 3'))
      .attr('stroke', 'rgba(0, 0, 0, 0.54)')
    svg.append('text')
      .attr('x', x(dashLine()) - 40)
      .attr('y', margin.top - 5)
      .text(`<${Math.trunc(firstGradeAfterBinnedGrade)}%`)
      .attr('font-size', '0.875rem')
      .attr('line-height', '1.46429em')
      .attr('text-anchor', 'start')
      .attr('fill', 'rgba(0, 0, 0, 0.54)')
  }
}

export default createHistogram
