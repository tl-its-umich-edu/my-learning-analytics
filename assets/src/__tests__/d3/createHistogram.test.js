/* global describe, test, expect */
import createHistogram from '../../components/d3/createHistogram'
import { histogramData } from '../testData/d3TestData'

describe('createHistogram', () => {
  test('should build a histogram', () => {
    const div = document.createElement('div')
    createHistogram({ data: histogramData, width: 1000, height: 500, el: div })
    expect(div).toMatchSnapshot()
  })
})
