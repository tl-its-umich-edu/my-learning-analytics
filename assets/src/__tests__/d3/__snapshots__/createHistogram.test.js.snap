// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createHistogram should build a histogram 1`] = `
<div>
  <svg
    height="450"
    width="940"
  >
    <rect
      fill="steelblue"
      height="0"
      width="16.6"
      x="41"
      y="420"
    >
      <text
        color="white"
        height="30"
        text-anchor="middle"
        x="8.8"
        y="10"
      />
    </rect>
    <rect
      fill="steelblue"
      height="0"
      width="16.6"
      x="58.6"
      y="420"
    >
      <text
        color="white"
        height="30"
        text-anchor="middle"
        x="8.8"
        y="10"
      />
    </rect>
    <rect
      fill="steelblue"
      height="0"
      width="16.599999999999994"
      x="76.2"
      y="420"
    >
      <text
        color="white"
        height="30"
        text-anchor="middle"
        x="8.8"
        y="10"
      />
    </rect>
    <rect
      fill="steelblue"
      height="0"
      width="16.60000000000001"
      x="93.8"
      y="420"
    >
      <text
        color="white"
        height="30"
        text-anchor="middle"
        x="8.8"
        y="10"
      />
    </rect>
    <rect
      fill="steelblue"
      height="0"
      width="16.599999999999994"
      x="111.4"
      y="420"
    >
      <text
        color="white"
        height="30"
        text-anchor="middle"
        x="8.8"
        y="10"
      />
    </rect>
    <rect
      fill="steelblue"
      height="0"
      width="16.599999999999994"
      x="129"
      y="420"
    >
      <text
        color="white"
        height="30"
        text-anchor="middle"
        x="8.8"
        y="10"
      />
    </rect>
    <rect
      fill="steelblue"
      height="0"
      width="16.600000000000023"
      x="146.6"
      y="420"
    >
      <text
        color="white"
        height="30"
        text-anchor="middle"
        x="8.8"
        y="10"
      />
    </rect>
    <rect
      fill="steelblue"
      height="0"
      width="16.599999999999994"
      x="164.20000000000002"
      y="420"
    >
      <text
        color="white"
        height="30"
        text-anchor="middle"
        x="8.8"
        y="10"
      />
    </rect>
    <rect
      fill="steelblue"
      height="0"
      width="16.599999999999994"
      x="181.8"
      y="420"
    >
      <text
        color="white"
        height="30"
        text-anchor="middle"
        x="8.8"
        y="10"
      />
    </rect>
    <rect
      fill="steelblue"
      height="0"
      width="16.599999999999994"
      x="199.4"
      y="420"
    >
      <text
        color="white"
        height="30"
        text-anchor="middle"
        x="8.8"
        y="10"
      />
    </rect>
    <rect
      fill="steelblue"
      height="0"
      width="16.599999999999994"
      x="217"
      y="420"
    >
      <text
        color="white"
        height="30"
        text-anchor="middle"
        x="8.8"
        y="10"
      />
    </rect>
    <rect
      fill="steelblue"
      height="0"
      width="16.599999999999994"
      x="234.6"
      y="420"
    >
      <text
        color="white"
        height="30"
        text-anchor="middle"
        x="8.8"
        y="10"
      />
    </rect>
    <rect
      fill="steelblue"
      height="0"
      width="16.600000000000023"
      x="252.2"
      y="420"
    >
      <text
        color="white"
        height="30"
        text-anchor="middle"
        x="8.8"
        y="10"
      />
    </rect>
    <rect
      fill="steelblue"
      height="0"
      width="16.600000000000023"
      x="269.8"
      y="420"
    >
      <text
        color="white"
        height="30"
        text-anchor="middle"
        x="8.8"
        y="10"
      />
    </rect>
    <rect
      fill="steelblue"
      height="0"
      width="16.599999999999966"
      x="287.40000000000003"
      y="420"
    >
      <text
        color="white"
        height="30"
        text-anchor="middle"
        x="8.8"
        y="10"
      />
    </rect>
    <rect
      fill="steelblue"
      height="0"
      width="16.600000000000023"
      x="305"
      y="420"
    >
      <text
        color="white"
        height="30"
        text-anchor="middle"
        x="8.8"
        y="10"
      />
    </rect>
    <rect
      fill="steelblue"
      height="0"
      width="16.600000000000023"
      x="322.6"
      y="420"
    >
      <text
        color="white"
        height="30"
        text-anchor="middle"
        x="8.8"
        y="10"
      />
    </rect>
    <rect
      fill="steelblue"
      height="400"
      width="16.599999999999966"
      x="340.20000000000005"
      y="20"
    >
      <text
        color="white"
        height="430"
        text-anchor="middle"
        x="8.8"
        y="10"
      />
    </rect>
    <rect
      fill="steelblue"
      height="0"
      width="16.599999999999966"
      x="357.8"
      y="420"
    >
      <text
        color="white"
        height="30"
        text-anchor="middle"
        x="8.8"
        y="10"
      />
    </rect>
    <rect
      fill="steelblue"
      height="0"
      width="16.600000000000023"
      x="375.4"
      y="420"
    >
      <text
        color="white"
        height="30"
        text-anchor="middle"
        x="8.8"
        y="10"
      />
    </rect>
    <rect
      fill="steelblue"
      height="200"
      width="16.599999999999966"
      x="393"
      y="220"
    >
      <text
        color="white"
        height="230"
        text-anchor="middle"
        x="8.8"
        y="10"
      />
    </rect>
    <rect
      fill="steelblue"
      height="0"
      width="16.600000000000023"
      x="410.59999999999997"
      y="420"
    >
      <text
        color="white"
        height="30"
        text-anchor="middle"
        x="8.8"
        y="10"
      />
    </rect>
    <rect
      fill="steelblue"
      height="0"
      width="16.600000000000023"
      x="428.2"
      y="420"
    >
      <text
        color="white"
        height="30"
        text-anchor="middle"
        x="8.8"
        y="10"
      />
    </rect>
    <rect
      fill="steelblue"
      height="0"
      width="16.599999999999966"
      x="445.8"
      y="420"
    >
      <text
        color="white"
        height="30"
        text-anchor="middle"
        x="8.8"
        y="10"
      />
    </rect>
    <rect
      fill="steelblue"
      height="0"
      width="16.600000000000023"
      x="463.4"
      y="420"
    >
      <text
        color="white"
        height="30"
        text-anchor="middle"
        x="8.8"
        y="10"
      />
    </rect>
    <rect
      fill="steelblue"
      height="0"
      width="16.600000000000023"
      x="481"
      y="420"
    >
      <text
        color="white"
        height="30"
        text-anchor="middle"
        x="8.8"
        y="10"
      />
    </rect>
    <rect
      fill="steelblue"
      height="0"
      width="16.600000000000023"
      x="498.6"
      y="420"
    >
      <text
        color="white"
        height="30"
        text-anchor="middle"
        x="8.8"
        y="10"
      />
    </rect>
    <rect
      fill="steelblue"
      height="200"
      width="16.600000000000023"
      x="516.2"
      y="220"
    >
      <text
        color="white"
        height="230"
        text-anchor="middle"
        x="8.8"
        y="10"
      />
    </rect>
    <rect
      fill="steelblue"
      height="0"
      width="16.59999999999991"
      x="533.8000000000001"
      y="420"
    >
      <text
        color="white"
        height="30"
        text-anchor="middle"
        x="8.8"
        y="10"
      />
    </rect>
    <rect
      fill="steelblue"
      height="0"
      width="16.600000000000023"
      x="551.4"
      y="420"
    >
      <text
        color="white"
        height="30"
        text-anchor="middle"
        x="8.8"
        y="10"
      />
    </rect>
    <rect
      fill="steelblue"
      height="400"
      width="16.600000000000023"
      x="569"
      y="20"
    >
      <text
        color="white"
        height="430"
        text-anchor="middle"
        x="8.8"
        y="10"
      />
    </rect>
    <rect
      fill="steelblue"
      height="0"
      width="16.600000000000023"
      x="586.6"
      y="420"
    >
      <text
        color="white"
        height="30"
        text-anchor="middle"
        x="8.8"
        y="10"
      />
    </rect>
    <rect
      fill="steelblue"
      height="0"
      width="16.600000000000023"
      x="604.2"
      y="420"
    >
      <text
        color="white"
        height="30"
        text-anchor="middle"
        x="8.8"
        y="10"
      />
    </rect>
    <rect
      fill="steelblue"
      height="0"
      width="16.600000000000023"
      x="621.8000000000001"
      y="420"
    >
      <text
        color="white"
        height="30"
        text-anchor="middle"
        x="8.8"
        y="10"
      />
    </rect>
    <rect
      fill="steelblue"
      height="0"
      width="16.59999999999991"
      x="639.4000000000001"
      y="420"
    >
      <text
        color="white"
        height="30"
        text-anchor="middle"
        x="8.8"
        y="10"
      />
    </rect>
    <rect
      fill="steelblue"
      height="0"
      width="16.600000000000023"
      x="657"
      y="420"
    >
      <text
        color="white"
        height="30"
        text-anchor="middle"
        x="8.8"
        y="10"
      />
    </rect>
    <rect
      fill="steelblue"
      height="0"
      width="16.600000000000023"
      x="674.6"
      y="420"
    >
      <text
        color="white"
        height="30"
        text-anchor="middle"
        x="8.8"
        y="10"
      />
    </rect>
    <rect
      fill="steelblue"
      height="200"
      width="16.59999999999991"
      x="692.2"
      y="220"
    >
      <text
        color="white"
        height="230"
        text-anchor="middle"
        x="8.8"
        y="10"
      />
    </rect>
    <rect
      fill="steelblue"
      height="0"
      width="16.600000000000023"
      x="709.8"
      y="420"
    >
      <text
        color="white"
        height="30"
        text-anchor="middle"
        x="8.8"
        y="10"
      />
    </rect>
    <rect
      fill="steelblue"
      height="0"
      width="16.600000000000023"
      x="727.4"
      y="420"
    >
      <text
        color="white"
        height="30"
        text-anchor="middle"
        x="8.8"
        y="10"
      />
    </rect>
    <rect
      fill="steelblue"
      height="0"
      width="16.59999999999991"
      x="745"
      y="420"
    >
      <text
        color="white"
        height="30"
        text-anchor="middle"
        x="8.8"
        y="10"
      />
    </rect>
    <rect
      fill="steelblue"
      height="200"
      width="16.600000000000023"
      x="762.5999999999999"
      y="220"
    >
      <text
        color="white"
        height="230"
        text-anchor="middle"
        x="8.8"
        y="10"
      />
    </rect>
    <rect
      fill="steelblue"
      height="0"
      width="16.600000000000023"
      x="780.1999999999999"
      y="420"
    >
      <text
        color="white"
        height="30"
        text-anchor="middle"
        x="8.8"
        y="10"
      />
    </rect>
    <rect
      fill="steelblue"
      height="0"
      width="16.600000000000023"
      x="797.8"
      y="420"
    >
      <text
        color="white"
        height="30"
        text-anchor="middle"
        x="8.8"
        y="10"
      />
    </rect>
    <rect
      fill="steelblue"
      height="0"
      width="16.600000000000023"
      x="815.4"
      y="420"
    >
      <text
        color="white"
        height="30"
        text-anchor="middle"
        x="8.8"
        y="10"
      />
    </rect>
    <rect
      fill="steelblue"
      height="0"
      width="16.600000000000023"
      x="833"
      y="420"
    >
      <text
        color="white"
        height="30"
        text-anchor="middle"
        x="8.8"
        y="10"
      />
    </rect>
    <rect
      fill="steelblue"
      height="200"
      width="16.59999999999991"
      x="850.6"
      y="220"
    >
      <text
        color="white"
        height="230"
        text-anchor="middle"
        x="8.8"
        y="10"
      />
    </rect>
    <rect
      fill="steelblue"
      height="0"
      width="16.600000000000023"
      x="868.1999999999999"
      y="420"
    >
      <text
        color="white"
        height="30"
        text-anchor="middle"
        x="8.8"
        y="10"
      />
    </rect>
    <rect
      fill="steelblue"
      height="200"
      width="16.600000000000023"
      x="885.8"
      y="220"
    >
      <text
        color="white"
        height="230"
        text-anchor="middle"
        x="8.8"
        y="10"
      />
    </rect>
    <rect
      fill="steelblue"
      height="0"
      width="16.600000000000023"
      x="903.4"
      y="420"
    >
      <text
        color="white"
        height="30"
        text-anchor="middle"
        x="8.8"
        y="10"
      />
    </rect>
    <rect
      fill="steelblue"
      height="200"
      width="0"
      x="921"
      y="220"
    >
      <text
        color="white"
        height="230"
        text-anchor="middle"
        x="8.8"
        y="10"
      />
    </rect>
    <g
      fill="none"
      font-family="sans-serif"
      font-size="10"
      text-anchor="middle"
      transform="translate(0, 420)"
    >
      <path
        class="domain"
        d="M40.5,0.5H920.5"
        stroke="currentColor"
      />
      <g
        class="tick"
        opacity="1"
        transform="translate(40.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          0%
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(84.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          5%
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(128.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          10%
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(172.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          15%
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(216.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          20%
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(260.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          25%
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(304.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          30%
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(348.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          35%
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(392.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          40%
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(436.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          45%
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(480.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          50%
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(524.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          55%
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(568.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          60%
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(612.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          65%
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(656.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          70%
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(700.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          75%
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(744.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          80%
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(788.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          85%
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(832.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          90%
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(876.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          95%
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(920.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          100%
        </text>
      </g>
      <text
        fill="#000"
        text-anchor="end"
        x="920"
        y="-4"
      />
    </g>
    <g
      class="axis"
      fill="none"
      font-family="sans-serif"
      font-size="10"
      text-anchor="end"
      transform="translate(40,0)"
    >
      <g
        class="tick"
        opacity="1"
        transform="translate(0,420.5)"
      >
        <line
          stroke="currentColor"
          x2="940"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-3"
        >
          0.0
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,320.5)"
      >
        <line
          stroke="currentColor"
          x2="940"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-3"
        >
          0.5
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,220.5)"
      >
        <line
          stroke="currentColor"
          x2="940"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-3"
        >
          1.0
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,120.5)"
      >
        <line
          stroke="currentColor"
          x2="940"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-3"
        >
          1.5
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,20.5)"
      >
        <line
          stroke="currentColor"
          x2="940"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-3"
        >
          2.0
        </text>
        <text
          dy="-4"
          fill="#000"
          text-anchor="start"
          x="4"
        />
      </g>
    </g>
  </svg>
</div>
`;
